### ISSUE TITLE
New trains enter the sim if the route is not clear

### ISSUE THREAD

**Issue by hklang10** [ts2/ts2-sim-server#48]
Currently, the sim will generate new trains even if the track up to the first signal is already occupied. 

This can be re-created in Liverpool Street by cancelling the inbound signals at Bow Junction and running the sim for a period of time.

**Comment by hklang10** [comment]
Additional issue: if the first signal in the sim is at danger, trains entering the sim do not reduce their speed quickly enough.  This causes the train to go-through the signal and stop in the next track item.

**Comment by npiganeau** [comment]
That's true, there is no safety when entering the area, trains just appear at the entry point with their entry speed. Any idea on how to solve this issue is welcome.

**Comment by hklang10** [comment]
here is some work in progress.....

hklang10 branch "arrivals"

Sorry this is messy because it also includes code for another pending PR.  I will need to sort this out before pushing so is not ready for release yet.

Basically I have done three things:
a) for inactive trains, the veryhighspeed action has been removed and they now have "unlimited visibility" of the first signal - this prevents the trains arriving at fast speeds.
b) all track sections between the inactive service and the first signal are checked, if there is a train present, the activate does not occur.  Note: I've wrapped this check in a mutex to limit the chance of trains arriving simultaneously, not sure if that is correct or not.
c) the activate() function has been re-ordered for a) and b) to work.

During testing, I am concerned because I am hitting occasional issues of the sim freezing - problem is I don't know if this is being caused by this branch, of if it is an existing issue.

**Comment by hklang10** [comment]
Ah ok, the maxspeed branch has been merged, let me re-work this code in another branch so it is clear what I'm currently proposing.

**Comment by hklang10** [comment]
OK here is the compare - changes now on my branch iss48.
I'm not pushing currently because of the sim freezing and I'm not sure if this push could be the reason.

https://github.com/ts2/ts2-sim-server/compare/master...hklang10:iss48

**Comment by npiganeau** [comment]
Interesting stuff here. Maybe you should start a WIP PR with this and your fix for sim loading so that we can review it together, even if it freezes. 

@sharpe49 also reported freeze issues in #43 , so maybe it is the same issue and perhaps caused by 7a87dd26cf45fbe03d2bb8dd8c7c81d34542eeba
